impeller/display_list/aiks_dl_blur_unittests.cc

```diff
@@ -0,0 +1,971 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "display_list/display_list.h"
+#include "display_list/dl_blend_mode.h"
+#include "display_list/dl_builder.h"
+#include "display_list/dl_color.h"
+#include "display_list/dl_paint.h"
+#include "display_list/dl_sampling_options.h"
+#include "display_list/dl_tile_mode.h"
+#include "display_list/effects/dl_color_filter.h"
+#include "display_list/effects/dl_color_source.h"
+#include "display_list/effects/dl_image_filter.h"
+#include "display_list/effects/dl_mask_filter.h"
+#include "flutter/impeller/aiks/aiks_unittests.h"
+
+#include "impeller/display_list/dl_image_impeller.h"
+#include "impeller/playground/widgets.h"
+#include "include/core/SkRRect.h"
+#include "include/core/SkRect.h"
+#include "third_party/imgui/imgui.h"
+
+////////////////////////////////////////////////////////////////////////////////
+// This is for tests of Canvas that are interested the results of rendering
+// blurs.
+////////////////////////////////////////////////////////////////////////////////
+
+namespace impeller {
+namespace testing {
+
+using namespace flutter;
+
+TEST_P(AiksTest, CanRenderMaskBlurHugeSigma) {
+  DisplayListBuilder builder;
+
+  DlPaint paint;
+  paint.setColor(DlColor::kGreen());
+  paint.setMaskFilter(DlBlurMaskFilter::Make(DlBlurStyle::kNormal, 99999));
+  builder.DrawCircle({400, 400}, 300, paint);
+  builder.Restore();
+
+  ASSERT_TRUE(OpenPlaygroundHere(builder.Build()));
+}
+
+TEST_P(AiksTest, CanRenderForegroundBlendWithMaskBlur) {
+  // This case triggers the ForegroundPorterDuffBlend path. The color filter
+  // should apply to the color only, and respect the alpha mask.
+  DisplayListBuilder builder;
+  builder.ClipRect(SkRect::MakeXYWH(100, 150, 400, 400));
+
+  DlPaint paint;
+  paint.setColor(DlColor::kWhite());
+
+  Scalar sigma = 1 + (20) / 2;
+  paint.setMaskFilter(DlBlurMaskFilter::Make(DlBlurStyle::kNormal, sigma));
+  paint.setColorFilter(
+      DlBlendColorFilter::Make(DlColor::kGreen(), DlBlendMode::kSrc));
+  builder.DrawCircle({400, 400}, 200, paint);
+  builder.Restore();
+
+  ASSERT_TRUE(OpenPlaygroundHere(builder.Build()));
+}
+
+TEST_P(AiksTest, CanRenderForegroundAdvancedBlendWithMaskBlur) {
+  // This case triggers the ForegroundAdvancedBlend path. The color filter
+  // should apply to the color only, and respect the alpha mask.
+  DisplayListBuilder builder;
+  builder.ClipRect(SkRect::MakeXYWH(100, 150, 400, 400));
+
+  DlPaint paint;
+  paint.setColor(
+      DlColor::RGBA(128.0f / 255.0f, 128.0f / 255.0f, 128.0f / 255.0f, 1.0f));
+
+  Scalar sigma = 1 + 20 / 2;
```

Comment:
Originally this test was converting `Radius(20)` to `Sigma` via Impeller's conversion utility. The conversion utility does [radius > 0 ? radius / kKernelRadiusPerSigma + 0.5f : 0.0f](https://github.com/flutter/engine/blob/76a1c64f4e63f03a8593ed0214f2d933ad211818/impeller/geometry/sigma.cc#L16), which doesn't match this conversion. I'd continue to use Impeller's Radius/Sigma tool when porting over all these tests to make sure they're behaving the same. You can just pull the `Scalar` out of them after converting.

Reply:
I actually used that originally and it was still wrong, so this was a shot in the dark change.

Reply:
or, I was trying to use the coversions:
```
Sigma sigma = Radius(..);
blur(sigma.sigma)l
```

Which hopefully does the same thing?

Reply:
Or I guess replace all these conversions with `radius / kKernelRadiusPerSigma + 0.5f` inline... both are kind of evil but I guess this would be further away from Impeller's internals.

I think I started the trend of authoring these tests in terms kernel radii, which was an oversight on my part. The radius->sigma conversion is a pure implementation detail subject to change.

Reply:
> or, I was trying to use the coversions:
> 
> ```
> Sigma sigma = Radius(..);
> blur(sigma.sigma)l
> ```
> 
> Which hopefully does the same thing?

Yes, I didn't see this message. Recommended reading for GitHub:

![image](https://github.com/user-attachments/assets/d51467af-c175-4925-ae7c-199444009896)

I would expect this to be correct.
```
Sigma sigma = Radius(20);
blur(sigma.sigma);
```